### PR TITLE
qt6: use makeScopeWithSplicing

### PR DESCRIPTION
**Pull request by Artturin**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/qt6-packages.nix [hunk 1]
```diff
@@ -45,4 +51,11 @@ in
     qt6Support = true;
     suffix = "qt6";
   };
-})))
+
+  } // lib.optionalAttrs pkgs.config.allowAliases {
+    # Convert to a throw on 03-01-2023 and backport the change.
```
Comment (Artturin) [inline pkgs/top-level/qt6-packages.nix:56]: We did a similar thing with qt5 https://github.com/NixOS/nixpkgs/pull/245905

**Comment by amjoseph-nixpkgs** [comment]
I'm having trouble testing this in my tree.

There's probably something in `master` that it needs, but I'm stuck on `staging` until it merges back.

As soon as the next `staging -> master` merge happens I will test this.

**Comment by Artturin** [comment]
> I'm having trouble testing this in my tree.
> 
> There's probably something in `master` that it needs, but I'm stuck on `staging` until it merges back.
> 
> As soon as #263535 merge happens I will rebase to `master` and test this.

I didn't build anything, only made sure it evals

master is in staging https://github.com/NixOS/nixpkgs/actions/workflows/periodic-merge-6h.yml

**Comment by amjoseph-nixpkgs** [comment]
> master is in staging https://github.com/NixOS/nixpkgs/actions/workflows/periodic-merge-6h.yml

Yeah but since I'm stuck on staging every rebase is a mass-rebuild.  So I don't rebase often.

Once staging (and the commits on staging that I currently need) merges back to master I can resume my usual rebase-every-few-days.

**Comment by Artturin** [comment]
> > master is in staging https://github.com/NixOS/nixpkgs/actions/workflows/periodic-merge-6h.yml
> 
> Yeah but since I'm stuck on staging every rebase is a mass-rebuild. So I don't rebase often.
> 
> Once staging (and the commits on staging that I currently need) merges back to master I can resume my usual rebase-every-few-days.
> 
> We've been in "staging will merge back to master in the next day or two" for like a week now. The merge always seems to be a day or two away, so I just keep waiting for it.

👍 

The reason -next hasn't been merged is because the build farm is starved of aarch64-linux and there was a necessary darwin fix merged from staging to -next.

The staging and infrastructure rooms on matrix have relevant discussions

File: pkgs/development/libraries/qt-6/default.nix [hunk 2]
```diff
@@ -163,26 +169,32 @@ let
...
-          propagatedBuildInputs = [ self.qtbase.dev ];
+          propagatedBuildInputs = [ qtbase.dev ];
           substitutions = {
             inherit debug;
             fix_qmake_libtool = ./hooks/fix-qmake-libtool.sh;
           };
-        } ./hooks/qmake-hook.sh;
+        } ./hooks/qmake-hook.sh) { };
+    } // lib.optionalAttrs config.allowAliases {
+      # Convert to a throw on 03-01-2023 and backport the change.
```
Comment (ShamrockLee) [inline pkgs/development/libraries/qt-6/default.nix:188]: Not sure when is exactly 03-01-2023, but isn't it already passed?

Comment (Artturin) [reply pkgs/development/libraries/qt-6/default.nix:188]: Meant 2024 :P

I have it in my calendar so no need to fix the typo

File: pkgs/top-level/qt6-packages.nix [hunk 3]
```diff
@@ -5,17 +5,27 @@
...
 
-(lib.makeScope pkgs.newScope ( self:
-
 let
-  callPackage = self.callPackage;
+  pkgs = __splicedPackages;
+  # qt6 set should not be pre-spliced to prevent spliced packages being a part of an unspliced set
+  # 'pkgsCross.aarch64-multiplatform.pkgsBuildTarget.targetPackages.qt6Packages.qtbase' should not have a `__spliced` but if qt6 is pre-spliced then it will have one.
+  # pkgsBuildHost == pkgs
+  qt6 = pkgsBuildHost.qt6;
```
Comment (doronbehar) [inline pkgs/top-level/qt6-packages.nix:20]: I have a question @Artturin : how come this workaround is not needed for `qt6Packages` in `pkgs/kde/default.nix`?

**Pull request merged**